### PR TITLE
Fix #91306 by deriving all access from a single *mut T

### LINKED ISSUE #91306: Miri detects a violation of stacked borrows in slice::sort_unstable with -Zmiri-track-raw-pointers
Similar to https://github.com/rust-lang/rust/issues/78749?

I tried this code:
```rust
fn main() {
    let mut data = [1, 0];
    data.sort_unstable();
}
```
`MIRIFLAGS="-Zmiri-track-raw-pointers" cargo +nightly miri run`

I expected to see this happen: Miri does not detect UB

Instead, this happened: Miri detects UB (experimentally)

```
error: Undefined Behavior: no item granting read access to tag <2552> at alloc1081 found in borrow stack.
    --> /home/ben/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/intrinsics.rs:2079:14
     |
2079 |     unsafe { copy_nonoverlapping(src, dst, count) }
     |              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ no item granting read access to tag <2552> at alloc1081 found in borrow stack.
     |
     = help: this indicates a potential bug in the program: it performed an invalid operation, but the rules it violated are still experimental
     = help: see https://github.com/rust-lang/unsafe-code-guidelines/blob/master/wip/stacked-borrows.md for further information
```

### Meta

`rustc --version --verbose`:
```
rustc 1.58.0-nightly (6d246f0c8 2021-11-26)
binary: rustc
commit-hash: 6d246f0c8d3063fea86abbb65a824362709541ba
commit-date: 2021-11-26
host: x86_64-unknown-linux-gnu
release: 1.58.0-nightly
LLVM version: 13.0.0
```

<details><summary>Backtrace</summary>
<p>

```
error: Undefined Behavior: no item granting read access to tag <2552> at alloc1081 found in borrow stack.
    --> /home/ben/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/intrinsics.rs:2079:14
     |
2079 |     unsafe { copy_nonoverlapping(src, dst, count) }
     |              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ no item granting read access to tag <2552> at alloc1081 found in borrow stack.
     |
     = help: this indicates a potential bug in the program: it performed an invalid operation, but the rules it violated are still experimental
     = help: see https://github.com/rust-lang/unsafe-code-guidelines/blob/master/wip/stacked-borrows.md for further information

     = note: inside `std::intrinsics::copy_nonoverlapping::<i32>` at /home/ben/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/intrinsics.rs:2079:14
     = note: inside `core::slice::sort::shift_tail::<i32, [closure@core::slice::<impl [i32]>::sort_unstable::{closure#0}]>` at /home/ben/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/slice/sort.rs:104:13
     = note: inside `core::slice::sort::insertion_sort::<i32, [closure@core::slice::<impl [i32]>::sort_unstable::{closure#0}]>` at /home/ben/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/slice/sort.rs:175:9
     = note: inside `core::slice::sort::recurse::<i32, [closure@core::slice::<impl [i32]>::sort_unstable::{closure#0}]>` at /home/ben/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/slice/sort.rs:734:13
     = note: inside `core::slice::sort::quicksort::<i32, [closure@core::slice::<impl [i32]>::sort_unstable::{closure#0}]>` at /home/ben/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/slice/sort.rs:815:5
     = note: inside `core::slice::<impl [i32]>::sort_unstable` at /home/ben/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/slice/mod.rs:2356:9
note: inside `main` at src/main.rs:3:5
    --> src/main.rs:3:5
     |
3    |     data.sort_unstable();
     |     ^^^^^^^^^^^^^^^^^^^^
     = note: inside `<fn() as std::ops::FnOnce<()>>::call_once - shim(fn())` at /home/ben/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/ops/function.rs:227:5
     = note: inside `std::sys_common::backtrace::__rust_begin_short_backtrace::<fn(), ()>` at /home/ben/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/sys_common/backtrace.rs:123:18
     = note: inside closure at /home/ben/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/rt.rs:145:18
     = note: inside `std::ops::function::impls::<impl std::ops::FnOnce<()> for &dyn std::ops::Fn() -> i32 + std::marker::Sync + std::panic::RefUnwindSafe>::call_once` at /home/ben/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/ops/function.rs:259:13
     = note: inside `std::panicking::r#try::do_call::<&dyn std::ops::Fn() -> i32 + std::marker::Sync + std::panic::RefUnwindSafe, i32>` at /home/ben/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/panicking.rs:406:40
     = note: inside `std::panicking::r#try::<i32, &dyn std::ops::Fn() -> i32 + std::marker::Sync + std::panic::RefUnwindSafe>` at /home/ben/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/panicking.rs:370:19
     = note: inside `std::panic::catch_unwind::<&dyn std::ops::Fn() -> i32 + std::marker::Sync + std::panic::RefUnwindSafe, i32>` at /home/ben/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/panic.rs:133:14
     = note: inside closure at /home/ben/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/rt.rs:128:48
     = note: inside `std::panicking::r#try::do_call::<[closure@std::rt::lang_start_internal::{closure#2}], isize>` at /home/ben/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/panicking.rs:406:40
     = note: inside `std::panicking::r#try::<isize, [closure@std::rt::lang_start_internal::{closure#2}]>` at /home/ben/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/panicking.rs:370:19
     = note: inside `std::panic::catch_unwind::<[closure@std::rt::lang_start_internal::{closure#2}], isize>` at /home/ben/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/panic.rs:133:14
     = note: inside `std::rt::lang_start_internal` at /home/ben/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/rt.rs:128:20
     = note: inside `std::rt::lang_start::<()>` at /home/ben/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/rt.rs:144:17

error: aborting due to previous error
```

</p>
</details>

### PR DESCRIPTION
**Pull request by saethlin**
Fixes #91306.

The previous code is invalid because the first argument to `copy_nonoverlapping` is invalidated by the mutable borrow taken out to construct the second argument.

I believe this patch fixes that, and this code should now pass Miri with `-Ztag-raw-pointers`, ~~but I'm currently stuck trying to run my reproducer with a this patched version of the standard library (alternatively, running Miri on the standard library tests itself would suffice).~~ Ralf walked me through this on Zulip.

I've also added fixes for 7 more problems other than those I reported. Most of them are easy to hit by calling sort_unstable on random arrays. I don't have reproducers for every change, but they seem pretty clear-cut to me. But I did only start learning stacked borrows 2 days ago so that might be a large dash of Dunning-Kruger.

### CODE REVIEW

**Comment by camelid** [comment]
This needs to be re-reviewed since the branch was updated.

@bors r-

File: library/core/src/slice/sort.rs [hunk 1]
```diff
@@ -101,15 +101,21 @@ where
             // into the slice.
             let mut tmp = mem::ManuallyDrop::new(ptr::read(v.get_unchecked(len - 1)));
             let mut hole = CopyOnDrop { src: &mut *tmp, dest: v.get_unchecked_mut(len - 2) };
-            ptr::copy_nonoverlapping(v.get_unchecked(len - 2), v.get_unchecked_mut(len - 1), 1);
+            let ptr = v.as_mut_ptr();
+            let src = (ptr as *const T).offset(len as isize - 2);
```
Comment (Mark-Simulacrum) [inline library/core/src/slice/sort.rs]: nit: `.add(len - 2)` seems better here, we don't want negative offsets here I assume?

File: library/core/src/slice/sort.rs [hunk 2]
```diff
@@ -101,15 +101,21 @@ where
...
+            let dst = ptr.offset(len as isize - 1);
+            ptr::copy_nonoverlapping(src, dst, 1);
```
Comment (Mark-Simulacrum) [inline library/core/src/slice/sort.rs]: If we're changing these anyway, generally `copy_to_nonoverlapping` seems a little better to me.

**Review by Mark-Simulacrum — CHANGES_REQUESTED** [review]
Would be good to squash commits as well.

**Comment by camelid** [comment]
I edited the PR description to mark this as auto-closing #91306 given the PR title.

**Comment by saethlin** [comment]
I'm very confused. The diff in GitHub doesn't look anything like what I have locally. Going to squash this down and see if the situation resolves...

File: library/core/src/slice/sort.rs (hunk 2 above)
Comment (saethlin) [reply library/core/src/slice/sort.rs]: I don't disagree, but there are number of `copy_nonoverlapping` calls in this file. Should I change all of them?

Comment (Mark-Simulacrum) [reply library/core/src/slice/sort.rs]: Let's leave it for now.

File: library/core/src/slice/sort.rs [hunk 3]
```diff
@@ -579,7 +581,8 @@ where
 
             // Swap the found pair of out-of-order elements.
             r -= 1;
-            ptr::swap(v.get_unchecked_mut(l), v.get_unchecked_mut(r));
+            let ptr = v.as_mut_ptr();
+            ptr::swap(ptr.offset(l as isize), ptr.offset(r as isize));
```
Comment (Mark-Simulacrum) [inline library/core/src/slice/sort.rs:585]: As noted before, this should be `.add` probably, right?

**Comment by rylev** [comment]
This seems likely to be noise as a result of the issue tracked in https://github.com/rust-lang/rustc-perf/issues/1126 -- marking as triaged, not something we should address directly.

@rustbot label +perf-regression-triaged

**Pull request merged**